### ISSUE TITLE
Logrotate Errors on Username/Groups with Spaces

### ISSUE THREAD

**Issue by msangster** [logrotate/logrotate#105]
(was originally submitted to https://bugzilla.redhat.com/show_bug.cgi?id=1417106)

# Description of problem:
If you use a space in a user/group name then logrotate will error out indicating that the usage was wrong.

# Version-Release number of selected component (if applicable):
I do not believe this is version related but in my test it was: logrotate-3.8.6-7.el7_2.x86_64

# How reproducible:
Every time

# Steps to Reproduce:
1. Create a logrotate configuration and use a group name with spaces in the "su" or "create" directives.
2. Run: `logrotate -d config`
3. Error output

# Actual results
```
cat /etc/logrotate.d/test
/var/log/test/*.log {
    su Test "Test Group"
    weekly
    rotate 10
    copytruncate
    delaycompress
    compress
    notifempty
    missingok
}

logrotate -d /etc/logrotate.d/test
reading config file /etc/logrotate.d/test
error: /etc/logrotate.d/test:2 extra arguments for su
error: found error in /var/log/test/*.log , skipping
removing last 1 log configs
Allocating hash table for state file, size 64 entries

Handling 0 logs
```

# Expected results:
Whilst spaces are not traditionally used in groups and users and indeed I am certain a lot of tools wouldn't accept it in the username. We have integration with Active Directory where it is acceptable. As such I would expect logrotate to handle the group name with spaces at least but preferably user and group.

# Additional info:
I did experiment with other possible styles of inputting the group.

su Test Test\ Group
error: /etc/logrotate.d/test:2 extra arguments for su

su Test "Test\ Group"
error: /etc/logrotate.d/test:2 extra arguments for su

su Test Test\040Group
error: /etc/logrotate.d/test:2 unknown group 'Test\040Group'

(1024 is the GID of "Test Group")
su Test 1024
error: /etc/logrotate.d/test:2 unknown group '1024'

**Comment by alavrentiev** [comment]
You can try using this patch: https://github.com/logrotate/logrotate/issues/133
The plus sign (+) is used in place of spaces in group names.  User names cannot have spaces, though.
HTH

**Comment by kdudka** [comment]
@alavrentiev In #133 you are proposing a one-line patch that fixes only a test-case whereas this bug is about logrotate not accepting user names and groups containing spaces while parsing the configuration files.

**Comment by alavrentiev** [comment]
You can use a plus sign (instead of a space) in group names in config files, and with the patch they will work correctly.

Anton Lavrentiev
Contractor NIH/NLM/NCBI

**Comment by kdudka** [comment]
Which patch?  The only patch I know of changes just the test-suite.  It does not change the behavior of logrotate itself.

**Comment by alavrentiev** [comment]
See my comment in:

https://github.com/logrotate/logrotate/issues/133

Anton Lavrentiev
Contractor NIH/NLM/NCBI

**Comment by kule** [comment]
Haven't done c in a long time but this seems to boil down to this line here: https://github.com/logrotate/logrotate/blob/602b80357390cda910aa1746b1cc27d34d7bbd68/config.c#L299

Would it be reasonable to change the sscanf call here so it's aware of a escaped spaces and/or quoted values?

Like I said haven't touched c in ages, but I can have a stab if that's reasonable.

**Comment by kdudka** [comment]
There are two calls to `sscanf()` in `readModeUidGid()`.  Both of them would need to be replaced by some more flexible parser.

**Comment by alavrentiev** [comment]
Hi, you seem to have forgotten to tailor actual test cases with your fix.  In `test/test-common.sh`:
```
genconfig() {
    input=test-config.$1.in
    output=test-config.$1
    user=$(id -u -n)
    group=$(id -g -n)
    rootgroup=$(id -g -n root)
    sed "s,&DIR&,$PWD,g" < $input | sed "s,&USER&,$user,g" | sed "s,&GROUP&,$group,g" | sed "s,&ROOTGROUP&,$rootgroup,g" > $output
    chmod go-w $output
    config_crc=$(${MD5SUM} $output)
}
```
If either `user` or `group` above end up containing spaces (as it could be the case in the test environment with AD for group names -- Cygwin, for example), the tests configs will fail to read (as the names would have to be quoted).
```
$ id -gn
Domain Users
```
Tests that fail due to that are:
13:
```
error: test-config.13:7 extra arguments for createolddir
```
31:
```
error: test-config.31:1 extra arguments for create
```
69:
```
error: test-config.69:7 extra arguments for createolddir
```

**Comment by cgzones** [comment]
Thanks, fix pending in #595.
